src/fdc/MSXtar.cc

```diff
@@ -101,9 +102,9 @@ void MSXtar::parseBootSector(const MSXBootSector& boot)
 		throw MSXException(
 			"Illegal number sectors per FAT: ", sectorsPerFat);
 	}
-	if (nbRootDirSectors == 0) { // TODO: check limits more accurately
+	if (nbRootDirSectors == 0 || nbRootDirSectors * DIR_ENTRIES_PER_SECTOR != boot.dirEntries) {
```

Comment:
Is the first part of the check `nbRootDirSectors == 0` still needed?

Hmm, when I think about it more. You check that `((boot.dirEntries / 16) * 16) == boot.dirEntries`. So you effectively check that it's a non-zero multiple of 16. Might be clearer to express it like that?

Reply:
Still needs a check for zero, no?

Reply:
You're right.
I think I would be less confused if it were written like: `dirEntries == 0 || (dirEntries % DIR_ENTRIES_PER_SECTOR) != 0`.

Reply:
I did some further investigation, and it is actually not illegal for the number of root dir entries not to be a multiple of 16. However in practice it always is, and the code is currently not equipped to handle it so better to fail early.

If the number of root dir entries is not a multiple of 16, it should be rounded up to determine the sector count, and the places where it scans the root directory should stop when the count is reached.

I put an item on my TODO list to support arbitrary root dir entry counts.

Reply:
OK.
It's also fine for me to not support such obscure corner cases. (Don't complicate the code for something that will never trigger in practice).

Reply:
Also I just read on Wikipedia:

> This value must be adjusted so that directory entries always consume full logical sectors, given that each [directory entry](https://en.wikipedia.org/wiki/Design_of_the_FAT_file_system#DIR) takes up 32 bytes. MS-DOS/PC DOS require this value to be a multiple of 16. The maximum value supported on floppy disks is 240, the maximum value supported by MS-DOS/PC DOS on hard disks is 512.

So even though MSX-DOS does support odd-numbered directory entry counts (which is what I based my previous remark on), other systems do have this restriction, so all the more reason not to bother.